### PR TITLE
Add tensor support for tranforms

### LINKED ISSUE #137: Tensor Support for Transforms
Hi, your library looks awesome, but as I have already implemented my own dataset class which does not respect the torchio ImagesDataset structure, I can not easily use the transforms you implemented on my data. It would be amazing if they could be simply used on pytorch tensors as well.  

In brief: applying torchio transforms and augmentation to a pytorch tensor
If I understood your code correctly, this would mainly involve rewriting the ```apply_transform``` of each transform class

Is this something you had in mind?

### PR DESCRIPTION
**Pull request by fepegar**
Resolves #137.

### CODE REVIEW

File: torchio/transforms/transform.py [hunk 1]
```diff
@@ -58,6 +76,33 @@ def parse_sample(sample: Subject) -> None:
...
+                name = f'channel_{i}'
+                path = tmpdir / f'{name}.nii'
+                write_image(channel_tensor, affine, path)
+                image = Image(path, INTENSITY)
+                subject_dict[name] = image
+            subject = Subject(subject_dict)
+            dataset = ImagesDataset([subject])
+            sample = dataset[0]  # load the images
+        return sample
+
```
[the remark this answers is not in the document]
Comment (fepegar) [reply torchio/transforms/transform.py:105]: That would definitely make more sense and be more efficient, but this was my first idea. Maybe a solution would be modifying `Image` to get data and affine, not just a path.

[the remark this answers is not in the document]
Comment (fepegar) [reply torchio/transforms/transform.py:100]: I modified `Image` and `Transform` to avoid writing to disk.

**Comment by MonsieurWave** [comment]
The tensor with channels is accepted now, but the channels are lost upon transformation
```python
In [39]: sample_input = torch.ones((4,30,30,30))                                                                                                                                                            

In [40]: from torchio import RandomFlip                                                                                                                                                                     

In [41]: rflip = RandomFlip()                                                                                                                                                                               

In [42]: trans_sample_input = rflip(sample_input)                                                                                                                                                           

In [43]: trans_sample_input.shape                                                                                                                                                                           
Out[43]: torch.Size([30, 30, 30])
```

**Comment by fepegar** [comment]
Not really. That shape is a `@property` of `Subject` that returns the shape of the first image in the sample. Your `trans_sample_input` is an instance of `Subject` which inherits from `dict`. You'll find your tensors in the keys `channel_0` to `channel_3`. Try `trans_sample_input['channel_0'][torchio.DATA].shape`.

**Comment by MonsieurWave** [comment]
I see. But wouldn't the more intuitive behaviour be that it returns a tensor instead of a subject object? It would integrate better with a pure pytorch syntax at least.

**Comment by fepegar** [comment]
You're right. Transforms now return tensors if tensors were given.

**Comment by fepegar** [comment]
I've messed up the commits trying to rebase, but the changes are hopefully still fine. I'll squash and merge if you're ok with the new behavior.

**Comment by MonsieurWave** [comment]
Looks good to me, it seems to work well on my data. 
Thanks a lot for taking the time to implement this!

**Comment by fepegar** [comment]
Thanks for reporting and testing, @MonsieurWave!

**Comment by fepegar** [comment]
@all-contributors please add @MonsieurWave for ideas

**Comment by fepegar** [comment]
@all-contributors please add @MonsieurWave for ideas

**Pull request merged**